Navigation: Show suggestions panel automatically when editing a link

jeryj opened the pull request:
Would close https://github.com/WordPress/gutenberg/issues/19647 if we go this route.

## What it fixes
1. Show the suggestions panel automatically when editing a link.
Before: 
![suggestions-panel-no-url-suggestion](https://user-images.githubusercontent.com/967608/74041450-467b7800-498b-11ea-9d80-e24acba44ab7.gif)

2. ~Fix suggestions reset when resetting from a URL-like search result~ Reset button removed in https://github.com/WordPress/gutenberg/pull/20052
Before:
![suggestions-not-reset-when-url-like-results](https://user-images.githubusercontent.com/967608/74041295-01efdc80-498b-11ea-823e-58a76a9e1d5d.gif)

3. ~Applying focus to the search input on reset (currently it gets reset to the top of the document )~ Reset button removed in https://github.com/WordPress/gutenberg/pull/20052
Before: 
![suggestions-reset-focus-loss](https://user-images.githubusercontent.com/967608/74041308-087e5400-498b-11ea-8359-721eb0dca2c6.gif)

## Description
- Adds an `onFocus` handler to `URLInput` so it can be used by the `LinkControl` `SearchInput` to automatically show suggestions when the search field is focused.
- ~Resets search suggestions when X is clicked or the field is deleted~
- ~Focuses search input when X is clicked~

## How has this been tested?
1. Manually
2. Integration Tests

## Screenshots 
![suggestions-reset-focus-fixed](https://user-images.githubusercontent.com/967608/74041346-1af88d80-498b-11ea-863c-2f51e9cb402c.gif)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

jeryj committed:
Trying to trigger suggestion search on focus when editing the link control.

jeryj commented:
@WunderBart @getdave I opened this draft PR to play around with. Feel free to take it over, abandon, or do whatever you'd like to it :)

It adds an `onFocus` handler to the `URLInput` that is just copy/pasted from the current `onChange`. That at least keeps them a little bit insulated from each other if we need them to diverge.

This implementation isn't an improvement though, IMO, as if you hit Enter, it will update the label to the link. I'd rather it bring up the matched Page than search the URL. It looks like it _used_ to search the label (ie About instead of localhost:8889/etc)

If we could go back to that implementation, this could possibly be good to go.

obenland commented:
>  It looks like it used to search the label (ie About instead of localhost:8889/etc) If we could go back to that implementation, this could possibly be good to go.

Agreed. Having the full url in the search input feels a little odd, I would have expected to see the link text.

obenland commented:
@apeatling Do you have an opinion on this change? 

It surfaces the lack of awareness that the link is actually an internal (page) resource and turns the entity into a normal absolute URL. Given that, does this PR feel like an improvement to you?

jeryj commented:
> It looks like it used to search the label (ie About instead of localhost:8889/etc)

I think I may have been wrong about this. It looks to me like the implementation of this PR is the same as it would be in `master`.  Testing it out again now, both implementations (`master` vs this branch) are identical other than this branch automatically searches whatever is placed in the input.

`master`:
![master-no-search-trigger-on-focus](https://user-images.githubusercontent.com/967608/73785069-3893e000-475c-11ea-8244-dc5711b5a7c1.gif)

This branch:
![search-result-on-focus](https://user-images.githubusercontent.com/967608/73785131-54978180-475c-11ea-9b14-257c02f22407.gif)

> It surfaces the lack of awareness that the link is actually an internal (page) resource and turns the entity into a normal absolute URL. 
I completely agree. We'd need this PR (or the like) in order to get the functionality that is desired (return the page result), but I'm not sure if that functionality should be within this PR. I'm not sure what the best thing to do here is.

obenland reviewed (approved):
You're right, I noticed that, too, while testing another PR.

apeatling commented:
I think this is an improvement, especially for the situation in #19647.

retrofox committed:
link-control: update main snapshot

jeryj committed:
Only run the onFocus updateSuggestions is there is an inputValue present

This ends up fixing almost all the failed tests.

jeryj committed:
Adds a reset method to the url-input

We needed a way to:
1. focus the input on reset
2. Reset the updateSuggestions

This method checks if there was a value and no longer has a value in order to perform the reset. Now it can fire programatically when a value is cleared.

jeryj committed:
Updated test to reflect state after clicking edit in the link control

jeryj committed:
Moved reset with initial suggestions to a different test

obenland commented:
@getdave When you get a chance, would you mind giving these test changes a final 👍?

getdave commented on packages/block-editor/src/components/link-control/test/index.js:
```diff
@@ -450,17 +448,52 @@ describe( 'Default search suggestions', () => {
+			'[role="listbox"] [role="option"]'
+		);
+
+		// Grab the reset button now it's available
```
As well as a testing for clicking on the reset UI I think we should test what happens when the user simply **deletes** the value in the input manually. Just trigger a change on the `input` with a empty string value.

jeryj replied:
Added a test for this. 👍

getdave replied:
Nice 👍

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -310,6 +326,13 @@ class URLInput extends Component {
+		// move focus to the input
+		this.inputRef.current.focus();
+		// reset our suggestions
+		this.updateSuggestions();
```
Do we need to call `this.updateSuggestions()` here given that it is called in `onFocus` or do you feel that `reset` shouldn't be aware of this fact?

jeryj replied:
The issue I was running into is that the suggestions were not returning to their initial state because there was a manual URL suggestion hanging around. You can see the same effect on master by typing a URL-like string then clicking the reset x.

jeryj replied:
I tested this out again, and it seems that we need to forcefully call `this.updateSuggestions()` to get it to work correctly.

The other option was to do `this.setState` manually to reset the suggestions, but I was worried about the impact of manually setting that.

getdave replied:
> we need to forcefully call this.updateSuggestions() to get it to work correctly.

Isn't that a little odd then? Because the `this.inputRef.current.focus();` above also ends up triggering `this.updateSuggestions` via the `onFocus` handler. 

I wonder whether it would be better to just manually reset the suggestions here rather than relying on a side effect of calling `updateSuggestions` without a value. At least that way we're being explicit about our intent (ie: if the input has no value on update then clear all suggestions).

I'd also value an opinion from @aduth here.

jeryj replied:
https://github.com/WordPress/gutenberg/pull/19850/files#diff-7f80e90a821240f277aae4a7d2d9e778R88-R91

The focus only fires `this.updateSuggestions` if there is an input value present. Loads of tests were failing by forcefully calling it anytime there was a focus.

jeryj replied:
Doing something like 
```
	reset() {
		// move focus to the input
		this.inputRef.current.focus();
		// reset our suggestions
		this.setState({
			suggestions: [],
			selectedSuggestion: null,
		});
	}
```

does work, but the tests I wrote fail. I can take a look at fixing those i you feel like this is a better way to go.

getdave replied:
This doesn't look relevant any more right @jeryj cos `reset` got removed?

jeryj replied:
Yup! Can be resolved IMO

jeryj replied:
I removed the entire `reset()` function just a bit ago

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -310,6 +326,13 @@ class URLInput extends Component {
+	reset() {
+		// move focus to the input
+		this.inputRef.current.focus();
+		// reset our suggestions
```
>  reset our suggestions

Minor: are we actually "resetting" though? Because calling with an empty string might trigger a `initialSuggestions` if the `showInitialSuggestions` prop is provided.

jeryj replied:
Maybe I’m confused about the update suggestions. I’m trying to return them to the initial suggestions if the field is empty and initial suggestions is on. When the cleared text was URL-like, the manual URL suggestion lingers around when it should be cleared.

jeryj replied:
Would you recommend editing the comment for clarification?

getdave replied:
Actually thinking about it, that's what happens in the `updateSuggestions` so maybe the clarification is fine. I was being OTT.

getdave replied:
I was being OTT.

getdave reviewed (commented):
Thanks for these changes @jery. Extra props for updating the tests 👍 

I wonder whether we also ought to have a test which explicitly tests the new `focus` behaviour you've added?

For example, testing that when the input receives a `focus` event then the appropriate results are shown. You could also test that the mocked fetch handler has been called (or not!) as required.

jeryj committed:
Test that 'Recently updated' label is present when resetting the UrlInput

jeryj committed:
Added test to check for manually deleting text from link control search input should show initial suggestions

jeryj committed:
Added line to existing test to make sure focus is on search input after reset

jeryj commented:
@getdave I've added a few more tests as recommended, including a check to make sure focus gets placed on the input after reset. Mind taking another look?

getdave reviewed (changes requested):
In general I feel this works but I have concerns about the implications of having to call `this.updateSuggestions` in the `reset` method.

Also as this component underpins all of the Link UI across the editor I'd recommend we get a 2nd opinion from @aduth before merging.

jeryj commented:
@aduth @getdave, As an update to make this PR a little easier to review, I updated the PR description as it is now addressing three different things.

I think the biggest one is the focus-loss after clicking the X to reset the input. It'd be great to get this in sooner than later, _especially_ since it is being used for all link searches now.

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -82,6 +84,11 @@ class URLInput extends Component {
 		}
+
+		// Reset the component to the starting state when things are cleared
+		if ( prevProps.value && ! value ) {
```
If you're doing this only to reset the input when the "RESET" button is clicked in LinkControl then you _could_ consider passing a ref into `URLInput` and then when `onReset` is called in `LinkControl` you could `.focus()` on the ref which would then be forwarded down to the `<input>` within `URLInput`.

jeryj replied:
It would be ideal to be able to run this reset only when the rest button is clicked over in `packages/block-editor/src/components/link-control/search-input.js`, but I wasn't able to get something there to work due to the lack of input reference. We may be over-engineering things if we use a forward ref though.

getdave replied:
Looks like reset was removed by @aduth so we can ditch this.

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -310,6 +326,13 @@ class URLInput extends Component {
 
+	reset() {
+		// move focus to the input
+		this.inputRef.current.focus();
```
See https://github.com/WordPress/gutenberg/pull/19850/files#r376489345

jeryj committed:
Removes the check for the presence of suggestions before showingInitialSuggestions.

This is unnecessary as we actually want these suggestions to show again when the field is emptied, and it will not be repopulated with this check in place if the text deleted had a URL-like suggestion

jeryj committed:
Removed call to updateSuggestions on reset(), as it will now get called again on componentDidUpdate due to previous commit that removes suggestion length check

jeryj committed:
Added clarifying comment to reset method as to why it calls a focus on the input

jeryj committed:
Added clarifying comment and reordered inputValue check to make it clearer what is happening and why

jeryj committed:
Don't rerun search suggestions on field focus if there are already suggestions present

jeryj committed:
Test for using the keyboard arrows to navigate initial suggestions when the search input is empty.

jeryj commented:
I think something is broken with the tests in the Link Control. I've written a test that should definitely fail. I found a bug after removing the check for `suggestions` in `shouldShowInitialSuggestions()`: https://github.com/WordPress/gutenberg/pull/19850/commits/faf6ba1e1f80446e341cf9810e4430de77855a54

When you use the down arrow to navigate into the links on a search, it does another search request via `componentDidUpdate`, then `shouldShowInitialSuggestions()` evaluates to `true` and runs the search again.
![cant use down arrow on empty search](https://user-images.githubusercontent.com/967608/74067808-aa1f9880-49bf-11ea-9c34-29ceaf3187eb.gif)

_The test I've written should fail_, but they don't. The only reason I can think of is that the test isn't actually waiting for another search result to finish. It seems only able to call a search one time in the tests.

Anyone with more testing experience have advice here?

jeryj committed:
Only attempt an update on suggestions if the input value has actually changed

jeryj committed:
Cleaned up comments

jeryj committed:
Updated snapshots after rebase

jeryj committed:
Cleaning up comments and code to make more sense given the changes from rebasing

jeryj commented:
This PR was pretty heavily effected by the removal of the X button in resetting the search (which is a good change): https://github.com/WordPress/gutenberg/pull/20052

PR description has been updated to match.

jeryj committed:
Remove reset() function as it is no longer relevant since the reset button was removed.

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -101,16 +107,15 @@ class URLInput extends Component {
 			__experimentalShowInitialSuggestions &&
-			! ( value && value.length ) &&
-			! ( suggestions && suggestions.length )
+			! ( value && value.length )
```
I'm not totally sure what effect removing `! ( suggestions && suggestions.length )` will have on other components that consume `URLInput`. That said `shouldShowInitialSuggestions` is a recent addition so I'm hopeful the impact would/should be fairly minimal.

jeryj replied:
I've reverted, as it was an issue related to using the reset button to update suggestions. As the reset button has been removed, this has been reverted.

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -79,7 +81,11 @@ class URLInput extends Component {
-		if ( this.shouldShowInitialSuggestions() ) {
+		// Only attempt an update on suggestions if the input value has actually changed.
+		if (
+			prevProps.value !== value &&
```
This seems logical. We don't want to trigger unecessarily.

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -212,6 +217,25 @@ class URLInput extends Component {
+		if (
+			inputValue &&
+			! this.props.disableSuggestions &&
+			! suggestions.length
```
What's the default value of `suggestions`? Will it _always_ have a `.length`.

jeryj replied:
It's an empty array.

jeryj replied:
I've made it a more robust check for `! ( suggestions && suggestions.length )`

getdave commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -212,6 +217,25 @@ class URLInput extends Component {
+		this.props.onFocus( inputValue );
+
+		// When opening the link editor, if there's a value present, we want to load the suggestions pane with the results for this input search value
+		// Don't re-run the suggestions on focus if there are already suggestions present (prevents searching again when tabbing between the input and buttons)
```
Great comment context 👍

getdave reviewed (approved):
Thanks for persisting with this 👍 

In general it looks good. I'm slightly concern about changing the conditionals around `shouldShowSuggestions()` but I'm a lot happier now that we're not manually forcing the current `focus`.

⚠️ I will caution here that this is a change effecting a main API so we should probably look for approval from a Core team member.

jeryj committed:
More robust check for suggestions presence

jeryj committed:
Reset the shouldShowInitialSuggestions back to previous state as the change was related to a reset() issue that is no longer relevant

jeryj commented:
> I'm slightly concern about changing the conditionals around `shouldShowSuggestions()`

That change was also related to fixing functionality around the `reset` button. As that has been removed entirely, I've reverted the `shouldShowSuggestions()` to its previous state.

obenland commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -212,6 +218,25 @@ class URLInput extends Component {
 	}
 
+	onFocus( event ) {
+		const { suggestions } = this.state;
```
Is there a reason `this.state` gets destructured but not `this.props`?

marekhrabe replied:
+1. I wouldn't really notice (and don't mind it) but `event` and `props` are both also used go get a single value from the object and are inlined

jeryj replied:
I had copy/pasted from `onChange` and then added in the `const { suggestions } = this.state;` because I needed access to the suggestions array. Then I didn't go back and refactor. That's the _why_ :) 

Do you think I should:
1. Refactor to have both `this.state` and `this.props` to be destructured at the top of the method
2. Refactor to remove `this.state` being destructured and use `this.state.suggestions` inline in the if statement below
3. Merge as is

obenland replied:
My vote is for 1. Not sure if there are coding guidelines for things like this yet.

jeryj replied:
That feels like the cleanest option to me as well.

obenland reviewed (approved):
Nice improvement over the existing experience. Tested a few posts functionally and works as expected.

marekhrabe reviewed (approved):
This is working well for me, based on testing instructions. Now the reset button is removed, it's much cleaner. I'd be happy to get this UX improvement in 👍

jeryj committed:
Refactored onfocus

aduth commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -212,6 +218,26 @@ class URLInput extends Component {
+
+		const inputValue = event.target.value;
+
+		onFocus( inputValue );
```
Why do we pass `inputValue` like this? Related to this, the API change should have been added to [the component's documentation](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/url-input/README.md), describing the value that's being passed.

Per my previous comment, I would expect that this component enforces that the `value` prop it is given will be the one shown in the input. Thus, the parent wouldn't need to receive the value argument, because it's the same as it was already providing in its rendered `value` prop.

jeryj replied:
> Why do we pass inputValue like this?

I had copied the `onChange` implementation. I'm still learning the ropes on Gutenberg, so I was trying to follow existing code as much as I could. 

> the API change should have been added to the component's documentation, describing the value that's being passed.

Thanks for pointing out the documentation - sorry I  did not update that! I'll get a PR in today to add it in.

jeryj replied:
I've created the PR here: https://github.com/WordPress/gutenberg/pull/20220

I did not add documentation as the `this.props.onFocus`  prop has been removed. If I am misunderstanding this and still need to add this to the component documentation, please let me know 👍

aduth commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -212,6 +218,26 @@ class URLInput extends Component {
+		const { suggestions } = this.state;
+		const { disableSuggestions, onFocus } = this.props;
+
+		const inputValue = event.target.value;
```
In what way would we expect this not to be the same as `this.props.value`?

jeryj replied:
Great point. I was following the `onChange` implementation. I can replace `event.target.value` with `this.props.value` (or, rather `value` and add `value` to `const { disableSuggestions, onFocus, value } = this.props`).

I will add that in to the same PR that the documentation update is in, unless you'd rather they be separated out.

aduth replied:
Ultimately, I question whether we need the argument _at all_.

With `onChange`, the value represents some _intent to change_ the value. With `onFocus`, I don't think we're wanting to effect any change in the value?

jeryj replied:
Completely agree. Thanks for walking me through this so clearly. Learning a lot from your comments! 💯
